File: python/day6.py
from enum import Enum

import numpy as np
import numpy.typing as npt


class Direction(Enum):
    Up = 0
    Down = 1
    Right = 2
    Left = 3


turn = (Direction.Right, Direction.Left, Direction.Down, Direction.Up)


def read_board(fname: str) -> tuple[npt.NDArray, tuple[int, int], Direction]:
    with open(fname, "r") as file:
        lines = [line.strip() for line in file.readlines()]

    pos = (-1, -1)
    dir = Direction.Down
    board = np.zeros((len(lines), len(lines[0])), np.int8)
    for i, line in enumerate(lines):
        for j, char in enumerate(line):
            if char == "#":
                board[i, j] = 1
            if char in {"<", ">", "^", "v"}:
                pos = (i, j)
                match char:
                    case "<":
                        dir = Direction.Left
                    case ">":
                        dir = Direction.Right
                    case "^":
                        dir = Direction.Up
                    case "v":
                        dir = Direction.Down

    return board, pos, dir


def hash_pos(pos: tuple[int, int], n_cols: int) -> int:
    return pos[0] * n_cols + pos[1]


def hash_pos_dir(pos: tuple[int, int], dir: Direction, n_cols: int) -> int:
    return pos[0] * n_cols * 4 + pos[1] * 4 + dir.value


def next_pos(pos: tuple[int, int], dir: Direction) -> tuple[int, int]:
    match dir:
        case Direction.Up:
            return (pos[0] - 1, pos[1])
        case Direction.Down:
            return (pos[0] + 1, pos[1])
        case Direction.Left:
            return (pos[0], pos[1] - 1)
        case Direction.Right:
            return (pos[0], pos[1] + 1)


def is_oob(pos: tuple[int, int], board: npt.NDArray) -> bool:
    if pos[0] < 0 or pos[0] >= board.shape[0]:
        return True
    if pos[1] < 0 or pos[1] >= board.shape[1]:
        return True

    return False


def check_loop(pos: tuple[int, int], dir: Direction, board: npt.NDArray) -> int:
    visited = set()
    while True:
        hpd = hash_pos_dir(pos, dir, board.shape[1])
        if hpd in visited:
            return 1
        visited.add(hpd)

        candidate = next_pos(pos, dir)
        if is_oob(candidate, board):
            return 0
        if board[candidate[0], candidate[1]] == 1:
            dir = turn[dir.value]
            continue

        pos = candidate


def day6a() -> None:
    board, pos, dir = read_board("../inputs/day6.txt")

    visited = set()
    while True:
        visited.add(hash_pos(pos, board.shape[1]))
        candidate = next_pos(pos, dir)
        if is_oob(candidate, board):
            break
        if board[candidate[0], candidate[1]] == 1:
            dir = turn[dir.value]
            continue

        pos = candidate

    print(len(visited))


def day6b() -> None:
    board, pos, dir = read_board("../inputs/day6.txt")

    num_loops = 0
    blocks_placed = set()
    while True:
        candidate = next_pos(pos, dir)
        if is_oob(candidate, board):
            break
        if board[candidate[0], candidate[1]] == 1:
            dir = turn[dir.value]
            continue

        hash_candidate = hash_pos(candidate, board.shape[0])
        if hash_candidate not in blocks_placed:
            blocks_placed.add(hash_candidate)
            board[candidate[0], candidate[1]] = 1
            num_loops += check_loop(pos, dir, board)
            board[candidate[0], candidate[1]] = 0

        pos = candidate

    print(num_loops)


day6b()

File: go/day06/day6.go
package day6

import (
	"fmt"

	"github.com/tadejsv/aoc-2024/utils"
)

type IntSet = map[int]bool

type Direction int

const (
	Right Direction = iota
	Left
	Up
	Down
)

var turn []Direction = []Direction{Down, Up, Right, Left}

func getBoard(fname string) ([][]int, [2]int, Direction) {
	lines, err := utils.ReadLines(fname)
	if err != nil {
		panic("Could not read lines")
	}

	var pos [2]int
	var dir Direction
	board := make([][]int, len(lines))
	for i, line := range lines {
		boardLine := make([]int, len(line))
		for j, char := range line {
			switch char {
			case '#':
				boardLine[j] = 1
			case '>':
				pos = [2]int{i, j}
				dir = Right
			case '<':
				pos = [2]int{i, j}
				dir = Left
			case 'v':
				pos = [2]int{i, j}
				dir = Down
			case '^':
				pos = [2]int{i, j}
				dir = Up
			}
		}
		board[i] = boardLine
	}

	return board, pos, dir
}

func hashPos(pos [2]int, nCols int) int {
	return pos[0]*nCols + pos[1]
}

func hashPosDir(pos [2]int, dir Direction, nCols int) int {
	return pos[0]*nCols*4 + pos[1]*4 + int(dir)
}

func nextPos(pos [2]int, dir Direction) [2]int {
	switch dir {
	case Up:
		pos = [2]int{pos[0] - 1, pos[1]}
	case Down:
		pos = [2]int{pos[0] + 1, pos[1]}
	case Left:
		pos = [2]int{pos[0], pos[1] - 1}
	case Right:
		pos = [2]int{pos[0], pos[1] + 1}
	}
	return pos
}

func isOOB(pos [2]int, board [][]int) bool {
	if pos[0] < 0 || pos[0] >= len(board) {
		return true
	}
	if pos[1] < 0 || pos[1] >= len(board[0]) {
		return true
	}
	return false
}

func checkLoop(pos [2]int, dir Direction, board [][]int) int {
	visited := make(IntSet)
	for {
		posHash := hashPosDir(pos, dir, len(board[0]))
		if _, ok := visited[posHash]; ok {
			return 1
		}
		visited[posHash] = true

		candidate := nextPos(pos, dir)
		if isOOB(candidate, board) {
			return 0
		}
		if board[candidate[0]][candidate[1]] == 1 {
			dir = turn[dir]
			continue
		}

		pos = candidate
	}
}

func Day6a() {
	board, pos, dir := getBoard("../inputs/day6.txt")

	visited := make(IntSet)
	for {
		visited[hashPos(pos, len(board[0]))] = true
		candidate := nextPos(pos, dir)
		if isOOB(candidate, board) {
			break
		}
		if board[candidate[0]][candidate[1]] == 1 {
			dir = turn[dir]
			continue
		}

		pos = candidate
	}

	fmt.Printf("%v\n", len(visited))
}

func Day6b() {
	board, pos, dir := getBoard("../inputs/day6.txt")

	numLoops := 0
	blocksPlaced := make(IntSet)
	for {
		candidate := nextPos(pos, dir)
		if isOOB(candidate, board) {
			break
		}
		if board[candidate[0]][candidate[1]] == 1 {
			dir = turn[dir]
			continue
		}

		blockHash := hashPos(candidate, len(board[0]))
		if _, ok := blocksPlaced[blockHash]; !ok {
			board[candidate[0]][candidate[1]] = 1
			numLoops += checkLoop(pos, dir, board)
			board[candidate[0]][candidate[1]] = 0
			blocksPlaced[blockHash] = true
		}

		pos = candidate
	}

	fmt.Printf("%v\n", numLoops)
}
